### PR TITLE
ID3v2: Report ambiguities for GRP1/TIT1 frames

### PR DESCRIPTION
**Pull request by uklotzde**
The changes depend on the definition of `__EXTRA_METADATA__`, which is always _undefined_ by default.

Due to re-ordering of the code the changes are somehow difficult to review. But now the code that depends on `__EXTRA_METADATA__` is better separated.

#### `__EXTRA_METADATA__` _undefined_ (default)

No functional changes. Only more detailed comments and a new `qInfo()` log when importing file tags to improve observability for the mess Apple created.

#### `__EXTRA_METADATA__` _defined_

A straightforward, unconditional mapping is used both for importing and exporting metadata. This will prevent inconsistencies and data loss when Mixxx supports the additional fields.

### CODE REVIEW

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you. The code looks good, but the comment can be improved.

File: src/track/taglib/trackmetadata_id3v2.cpp [hunk 1]
```diff
@@ -743,33 +743,48 @@ void importTrackMetadataFromTag(
...
     // https://discussions.apple.com/thread/7900430
     // http://blog.jthink.net/2016/11/the-reason-why-is-grouping-field-no.html
-    if (tag.frameListMap().contains("GRP1")) {
-        // New grouping/work mapping
-        const TagLib::ID3v2::FrameList appleGroupingFrames = tag.frameListMap()["GRP1"];
-        if (!appleGroupingFrames.isEmpty()) {
-            pTrackMetadata->refTrackInfo().setGrouping(
-                    firstNonEmptyFrameToQString(appleGroupingFrames));
-        }
+    const TagLib::ID3v2::FrameList traditionalGroupingFrames = tag.frameListMap()["TIT1"];
```
[the remark this answers is not in the document]
Comment (uklotzde) [reply src/track/taglib/trackmetadata_id3v2.cpp:746]: No. I am not going to discuss Apple's bullshit decision. The code and the comments contain enough information too look this up whenever needed.

File: src/track/taglib/trackmetadata_id3v2.cpp [hunk 2]
```diff
@@ -743,33 +743,48 @@ void importTrackMetadataFromTag(
...
-        const TagLib::ID3v2::FrameList workFrames = tag.frameListMap()["TIT1"];
-        if (!workFrames.isEmpty()) {
-            pTrackMetadata->refTrackInfo().setWork(
-                    firstNonEmptyFrameToQString(workFrames));
-        }
-        const TagLib::ID3v2::FrameList movementFrames = tag.frameListMap()["MVNM"];
-        if (!movementFrames.isEmpty()) {
-            pTrackMetadata->refTrackInfo().setMovement(
-                    firstNonEmptyFrameToQString(movementFrames));
+    // Use new grouping/work mapping unconditionally
```
[the remark this answers is not in the document]
Comment (uklotzde) [reply src/track/taglib/trackmetadata_id3v2.cpp]: https://github.com/mixxxdj/mixxx/pull/4812/commits/6e8caa56f54de8c7c8fd72d1ecaf5f4b8978a52c

File: src/track/taglib/trackmetadata_id3v2.cpp [hunk 3]
```diff
@@ -743,33 +743,48 @@ void importTrackMetadataFromTag(
...
+    const TagLib::ID3v2::FrameList movementFrames = tag.frameListMap()["MVNM"];
+    if (!movementFrames.isEmpty()) {
+        pTrackMetadata->refTrackInfo().setMovement(
+                firstNonEmptyFrameToQString(movementFrames));
+    }
+#else  // __EXTRA_METADATA__
+    // Use a conditional, content-sensitive strategy for determining which
+    // frames might contain the grouping metadata. If any "GRP1" frames are
+    // present in the file then these are preferred, even if their content
+    // is empty.
```
[the remark this answers is not in the document]
Comment (uklotzde) [reply src/track/taglib/trackmetadata_id3v2.cpp]: https://github.com/mixxxdj/mixxx/pull/4812/commits/6e8caa56f54de8c7c8fd72d1ecaf5f4b8978a52c

File: src/track/taglib/trackmetadata_id3v2.cpp [hunk 4]
```diff
@@ -1134,42 +1149,38 @@ bool exportTrackMetadataIntoTag(TagLib::ID3v2::Tag* pTag,
...
-    if (
 #if defined(__EXTRA_METADATA__)
-            !trackMetadata.getTrackInfo().getWork().isNull() ||
-            !trackMetadata.getTrackInfo().getMovement().isNull() ||
-#endif // __EXTRA_METADATA__
-            pTag->frameListMap().contains("GRP1")) {
-        // New grouping/work/movement mapping if properties for classical
-        // music are available or if the GRP1 frame is already present in
-        // the file.
+    // Use new grouping/work mapping unconditionally
```
[the remark this answers is not in the document]
Comment (uklotzde) [reply src/track/taglib/trackmetadata_id3v2.cpp]: https://github.com/mixxxdj/mixxx/pull/4812/commits/6e8caa56f54de8c7c8fd72d1ecaf5f4b8978a52c

**Comment by uklotzde** [comment]
Sigh. Looks like I am becoming Mixxx's top comment author instead of contributing any code :/

**Review by daschuer — APPROVED** [review]
Thank you. LGTM

**Pull request merged**